lib/stdlib/src/io_ansi.erl

```diff
@@ -0,0 +1,2099 @@
+    IsTerminal =
+        case whereis(user_drv) =/= self() of
+            true ->
+                case io:getopts(Device) of
+                    {error, _} -> false;
+                    Opts ->
+                        proplists:get_value(terminal, Opts, false)
+                end;
+            false ->
+                %% if called from within the user_drv process, we only check if
+                %% stdin/stdout are TTYs in order to avoid deadlocks
+                prim_tty:isatty(stdin) =:= true andalso prim_tty:isatty(stdout) =:= true
+        end,
+    IsSmartTerminal =
+        case os:type() of
+            {win32, _} -> true;
+            _ ->
+                try
+                    prim_tty:tigetstr("sgr0") =/= false
+                catch error:badarg ->
+                    false
+                end
+        end,
+    IsTerminal andalso IsSmartTerminal.
+
+-doc """
+Scan the string for virtial terminal sequences.
+
+The recognized VTSs will be converted into the corresponding `t:vts/0`.
+
+If you intend to parse arrow keys it is recommended that you first set the terminal in
+application mode by using `keypad_transmit_mode/0`. This will make it easier for
+`m:io_ansi` to correctly detect arrow keys.
+
+Any unrecognized [control sequence introducers](https://en.wikipedia.org/wiki/ANSI_escape_code#Control_Sequence_Introducer_commands),
+will be placed in a tuple tagged with `csi`.
+
+Example:
+
+```erlang
+1> io_ansi:scan("\eOA").
+[kcursor_up]
+2> io_ansi:scan("\eOB").
+[kcursor_down]
+3> io_ansi:scan(io_ansi:format([bold, "text"])).
+[bold, ~"text", reset]
+4> io_ansi:scan(io_ansi:format([{cursor, 0, 0}])).
+[{csi, ~"\e[1;1H"}, reset]
+```
+
+""".
+-spec scan(unicode:chardata()) -> [unicode:unicode_binary() | vts() | {csi, unicode:unicode_binary()}].
+scan(Data) ->
+    scan_binary(unicode:characters_to_binary(Data), <<>>, []).
+
+scan_binary(<<CSI/utf8, R/binary>> = Data, Bin, Acc) when CSI =:= $\e; CSI =:= 155 ->
+    case lookup_vts(Data) of
+        undefined ->
+            case re:run(Data, prim_tty:ansi_regexp(), [unicode]) of
+                {match, [{0, N}]} ->
+                    <<Ansi:N/binary, AnsiRest/binary>> = Data,
+                    scan_binary(AnsiRest, <<>>, [{csi, Ansi}, Bin | Acc]);
+                nomatch ->
+                        scan_binary(R, <<Bin/binary, CSI/utf8>>, Acc)
+            end;
+        {Code, _, Rest} ->
+            scan_binary(Rest, <<>>, [Code, Bin | Acc])
+    end;
+scan_binary(<<C/utf8, R/binary>>, Bin, Acc) ->
+    scan_binary(R, <<Bin/binary, C/utf8>>, Acc);
+scan_binary(<<>>, Bin, Acc) ->
+    [NonEmpty || NonEmpty <- lists:reverse([Bin | Acc]), NonEmpty =/= <<>>]. 
+
+lookup_vts(Data) ->
+    try
+        [case Data of
+                <<Value:(byte_size(Value))/binary, Rest/binary>> ->
+                    throw({Key, Value, Rest});
+                _ ->
+                    ok
+            end || Key := Values <- get_vts_mappings(),
+             Value <- Values],
```

Comment:
This code is un-parseable for me :-) 
Re-write with a fun() in the map-comprehension or something?